File: compare-sort-algorithms/src/counting_sort.c
#include <stdio.h>
#include <stdlib.h>
#include <time.h>
#include <stdbool.h>
#include <string.h>
#include <limits.h>

/**
 * Counting Sort implementation.
 * A non-comparison based sorting algorithm that works by counting the number of objects
 * that have each distinct key value, and using that information to determine their positions.
 * Efficient for small ranges of integers.
 */
void countingSort(int arr[], int n) {
    // Find the maximum value in the array
    int max = arr[0];
    for (int i = 1; i < n; i++) {
        if (arr[i] > max) {
            max = arr[i];
        }
    }
    
    // Find the minimum value in the array
    int min = arr[0];
    for (int i = 1; i < n; i++) {
        if (arr[i] < min) {
            min = arr[i];
        }
    }
    
    int range = max - min + 1;
    
    // Create count and output arrays
    int* count = (int*)calloc(range, sizeof(int));
    int* output = (int*)malloc(n * sizeof(int));
    
    if (count == NULL || output == NULL) {
        printf("Memory allocation failed\n");
        if (count) free(count);
        if (output) free(output);
        exit(1);
    }
    
    // Store the count of each element
    for (int i = 0; i < n; i++) {
        count[arr[i] - min]++;
    }
    
    // Store the cumulative count
    for (int i = 1; i < range; i++) {
        count[i] += count[i - 1];
    }
    
    // Build the output array
    for (int i = n - 1; i >= 0; i--) {
        output[count[arr[i] - min] - 1] = arr[i];
        count[arr[i] - min]--;
    }
    
    // Copy the output array to the input array
    memcpy(arr, output, n * sizeof(int));
    
    free(count);
    free(output);
}

/**
 * Reads numbers from file and returns as array.
 */
int* readNumbersFromFile(const char* filename, int* size) {
    FILE* file = fopen(filename, "r");
    if (file == NULL) {
        printf("Error: %s not found. Please run generate_data.py first.\n", filename);
        exit(1);
    }
    
    // Count the number of lines in the file
    int count = 0;
    char line[20]; // Assuming numbers won't exceed 20 characters
    
    while (fgets(line, sizeof(line), file)) {
        count++;
    }
    
    // Rewind file to beginning
    rewind(file);
    
    // Allocate memory for the numbers
    int* numbers = (int*)malloc(count * sizeof(int));
    if (numbers == NULL) {
        printf("Error: Memory allocation failed.\n");
        fclose(file);
        exit(1);
    }
    
    // Read numbers from file into array
    int i = 0;
    while (i < count && fgets(line, sizeof(line), file)) {
        numbers[i++] = atoi(line);
    }
    
    fclose(file);
    *size = count;
    return numbers;
}

/**
 * Checks if array is sorted.
 */
bool isSorted(int arr[], int size) {
    for (int i = 0; i < size - 1; i++) {
        if (arr[i] > arr[i + 1]) {
            return false;
        }
    }
    return true;
}

int main(int argc, char* argv[]) {
    printf("C Counting Sort Performance Test\n");
    printf("================================\n");
    
    // Get input and output filenames from command line arguments
    const char* filename = argc > 1 ? argv[1] : "random_list.txt";
    const char* result_filename = argc > 2 ? argv[2] : "results_c.txt";
    
    printf("Using dataset: %s\n", filename);
    
    // Read data from file
    printf("Reading data from file...\n");
    int data_size;
    int* data = readNumbersFromFile(filename, &data_size);
    printf("Data size: %d integers\n", data_size);
    
    // Create a copy for sorting (to preserve original)
    int* data_copy = (int*)malloc(data_size * sizeof(int));
    if (data_copy == NULL) {
        printf("Error: Memory allocation failed for data copy\n");
        free(data);
        exit(1);
    }
    memcpy(data_copy, data, data_size * sizeof(int));
    
    // Measure sorting time
    printf("Starting Counting Sort...\n");
    clock_t start_time = clock();
    countingSort(data_copy, data_size);
    clock_t end_time = clock();
    
    double execution_time = ((double)(end_time - start_time)) / CLOCKS_PER_SEC;
    
    // Verify the array is sorted
    bool sorted = isSorted(data_copy, data_size);
    
    // Results
    printf("Sorting completed: %s\n", sorted ? "SUCCESS" : "FAILED");
    printf("Execution time: %.6f seconds\n", execution_time);
    printf("Elements per second: %.0f\n", data_size / execution_time);
    
    // Save results to file
    FILE* result_file = fopen(result_filename, "w");
    if (result_file == NULL) {
        printf("Error: Could not create results file\n");
        free(data);
        free(data_copy);
        exit(1);
    }
    
    fprintf(result_file, "C Counting Sort Results\n");
    fprintf(result_file, "Data size: %d\n", data_size);
    fprintf(result_file, "Execution time: %.6f seconds\n", execution_time);
    fprintf(result_file, "Elements per second: %.0f\n", data_size / execution_time);
    fprintf(result_file, "Sorted correctly: %s\n", sorted ? "true" : "false");
    
    fclose(result_file);
    printf("Results saved to %s\n", result_filename);
    
    // Free memory
    free(data);
    free(data_copy);
    
    return 0;
}

File: compare-sort-algorithms/src/counting_sort.rs
use std::fs::File;
use std::io::{self, BufRead, BufReader, Write};
use std::time::Instant;
use std::env;

/// Counting Sort implementation in Rust.
/// A non-comparison based sorting algorithm that works by counting the number of objects
/// that have each distinct key value, and using that information to determine their positions.
/// Efficient for small ranges of integers.
fn counting_sort(arr: &mut [i32]) {
    if arr.is_empty() {
        return;
    }

    // Find the maximum and minimum values in the array
    let max = *arr.iter().max().unwrap();
    let min = *arr.iter().min().unwrap();
    let range = (max - min + 1) as usize;

    // Create count array and initialize with zeros
    let mut count = vec![0; range];
    let mut output = vec![0; arr.len()];

    // Store the count of each element
    for &x in arr.iter() {
        count[(x - min) as usize] += 1;
    }

    // Store the cumulative count
    for i in 1..range {
        count[i] += count[i - 1];
    }

    // Build the output array
    for &x in arr.iter().rev() {
        let index = (x - min) as usize;
        output[count[index] - 1] = x;
        count[index] -= 1;
    }

    // Copy the output array to the input array
    arr.copy_from_slice(&output);
}

/// Reads numbers from file and returns as a vector.
fn read_numbers_from_file(filename: &str) -> io::Result<Vec<i32>> {
    let file = File::open(filename)?;
    let reader = BufReader::new(file);
    
    let numbers: Result<Vec<i32>, _> = reader
        .lines()
        .map(|line| line?.trim().parse::<i32>().map_err(|e| io::Error::new(io::ErrorKind::InvalidData, e)))
        .collect();
    
    numbers
}

/// Verifies if the array is sorted.
fn is_sorted<T: Ord>(arr: &[T]) -> bool {
    arr.windows(2).all(|window| window[0] <= window[1])
}

fn main() -> io::Result<()> {
    println!("Rust Counting Sort Performance Test");
    println!("==================================");
    
    // Parse command line arguments
    let args: Vec<String> = env::args().collect();
    let filename = args.get(1).map_or("random_list.txt", |s| s.as_str());
    let result_filename = args.get(2).map_or("results_rust.txt", |s| s.as_str());
    
    println!("Using dataset: {}", filename);
    
    // Read data from file
    println!("Reading data from file...");
    let data = read_numbers_from_file(filename)?;
    println!("Data size: {} integers", data.len());
    
    // Create a copy for sorting (to preserve original)
    let mut data_copy = data.clone();
    
    // Measure sorting time
    println!("Starting Counting Sort...");
    let start = Instant::now();
    counting_sort(&mut data_copy);
    let duration = start.elapsed();
    
    let execution_time = duration.as_secs_f64();
    
    // Verify the array is sorted
    let sorted = is_sorted(&data_copy);
    
    // Results
    println!("Sorting completed: {}", if sorted { "SUCCESS" } else { "FAILED" });
    println!("Execution time: {:.6} seconds", execution_time);
    println!("Elements per second: {:.0}", data.len() as f64 / execution_time);
    
    // Save results to file
    let mut result_file = File::create(result_filename)?;
    writeln!(result_file, "Rust Counting Sort Results")?;
    writeln!(result_file, "Data size: {}", data.len())?;
    writeln!(result_file, "Execution time: {:.6} seconds", execution_time)?;
    writeln!(result_file, "Elements per second: {:.0}", data.len() as f64 / execution_time)?;
    writeln!(result_file, "Sorted correctly: {}", sorted)?;
    
    println!("Results saved to {}", result_filename);
    
    Ok(())
} 